expr_checker.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import ast
+from copy import copy
+
+
+_safe_type = None
+_readonly = None
+
+
+def check_type(obj):
+    if type(obj) in {list, tuple, set}:
+        list(map(check_type, obj))
+        return obj
+    elif type(obj) not in _safe_type:
+        raise ValueError(f"safe_eval didn't like {obj}")
+    else:
+        return obj
+
+
+def test(func, *args, **kwargs):
+    list(map(check_type, kwargs.values()))
+    list(map(check_type, args))
+
+    return check_type(func(*args, **kwargs))
+
+
+def check_ro(obj):
+    if type(obj) in _readonly:
+        return copy(obj)
+    else:
+        return obj
+
+
+class NodeChecker(ast.NodeTransformer):
+    def __init__(self, check_fn, allowed_attr):
+        self.check_fn = check_fn.__name__
+        self.allowed_attr = allowed_attr
+
+        super().__init__()
+
+    def visit_Call(self, node):
+        list(map(self.visit, node.args))
+
+        if isinstance(node.func, ast.Attribute):
+            if node.func.attr not in self.allowed_attr:
+                raise ValueError(
+                    f"safe_eval doesn't allow you to read {node.func.attr}")
+
+            name = copy(node.func.value)
+            node.func.value.__class__ = ast.Call
+            node.func.value.func = ast.Name("check_ro", ctx=ast.Load())
+            node.func.value.args = [name]
+            node.func.value.keywords = []
+        
+        else:
+            self.generic_visit(node.func)            
+            
+        return ast.Call(
+            func=ast.Name(self.check_fn, ctx=ast.Load()),
+            args=[node.func] + node.args,
+            keywords=node.keywords
+        )
+
+    def visit_FunctionDef(self, node):
+        if node.name in [self.check_fn, "check_ro", "check_type"]:
+            raise NameError(f"safe_eval: {node.name} is a reserved name")
+        
+        return node
+
+    def visit_Name(self, node):
+        if node.id in [self.check_fn, "check_ro", "check_type"]:
+            raise NameError(f"safe_eval: {node.id} is a reserved name")
+
+        return node
+
+    def visit_Attribute(self, node):
+        if node.attr not in self.allowed_attr:
+            raise ValueError(
+                f"safe_eval doesn't allow you to read {node.attr}")
+
+        if isinstance(node.ctx, ast.Load):
+            return ast.Call(
+                func=ast.Name("check_type", ctx=ast.Load()),
+                args=[node],
+                keywords=[]
+            )
+
+        else:
+            return ast.Call(
+                func=ast.Name("check_ro", ctx=ast.Load()),
+                args=[node.value],
+                keywords=[]
+            )
+
+    def visit_Assign(self, node):
+        list(map(self.visit, node.targets))
```

Comment:
isn't a solution to make `NodeChecker` explore all nodes recursively without needing to call this manually ? 

I think the right way to do it is to do (to double check)
```
node = self.generic_visit(node)
```

Can you update all method `def visit_Xxx` to call this ?

Reply:
I tried it with `visit_call` and the code breaks

Reply:
Why ?

Reply:
Ok, now it decided to work, nvm :joy:

Reply:
Well no it doesn't work, I made a mistake

imagine I do 

```python
x = 4
```

It'll become

```python
some_check(x) = some_check(4)
```

Reply:
```python
Module(
    body=[
        Assign(
            targets=[
                Name(id='x', ctx=Store())],
            value=Constant(value=4))],
    type_ignores=[])
```

```python
def visit_Name(self, node):
    if node.id in self.reserved_name:
        raise NameError(f"safe_eval: {node.id} is a reserved name")
return node
```

But here the left part is `ast.Name`, and the verification is not done at runtime for this type of node but at "compilation" (so you should not have `some_check(x)` but only `x`), am I missing something ?

Reply:
![screenshot_spect](https://user-images.githubusercontent.com/54316226/159480562-7b5df912-7f28-4348-b0bc-f11c263e151f.png)

I just did that, run your tests, and everything passed

Reply:
and for ast.Call, i just added those types in the safe_types list: 
```
types.BuiltinFunctionType, types.BuiltinMethodType, types.MethodType, types.MethodWrapperType, types.MethodDescriptorType, types.MethodWrapperType, types.ClassMethodDescriptorType
```

and do
![screenshot_spect](https://user-images.githubusercontent.com/54316226/159481757-e9b448b5-f2ed-437d-8b5a-7c19cc3fd8e0.png)

and every tests passed...